rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,481 @@
+tags. These can be used for matching labels and extracting parameters, as well
+as generating concrete label instances given a parameter value.
+
+```proto
+
+// Template for generating labels given parameter values, or matching labels
+// and extracting parameter values.
+message LabelTemplate {
+  repeated TagTemplate confidentiality_tags = 1;
+  repeated TagTemplate integrity_tags = 2;
+}
+
+// Template for generating a user-specific gRPC tag given a user ID, or for
+// extracting a user ID parameter from a matched gRPC tag.
+message GrpcTagTemplate {
+  string user_id_parameter_name = 1;
+}
+
+
+// Templated tags for supporting templated labels.
+message TagTemplate {
+  oneof tag {
+    GrpcTag grpc_tag = 1;
+    GrpcTagTemplate grpc_tag_template = 1;
+    WebAssemblyModuleTag web_assembly_module_tag = 2;
+    WebAssemblyModuleSignatureTag web_assembly_module_signature_tag = 3;
+    TlsEndpointTag tls_endpoint_tag = 4;
+  }
+}
+
+```
+
+As an example, the following can be used to define a label template matching a
+user-specific gRPC confidentiality tag and extracting the byte-array content of
+the tag as a parameter named user_id:
+
+```toml
+
+[[label_template.confidentiality_tags]]
+[label_template.confidentiality_tags.tag.grpc_tag_template]
+user_id_parameter_name = "user_id"
+
+```
+
+Note: I am unsure about the exact TOML representation for this.
+
+Question: In general TOML is not great for deeply nested object hierarchies, so
+we should perhaps consider a different serialisation format. Prost does not
+support text-serialised protos. JSON is an option, but not that great for human
+readable config. YAML can handle the complex object hierarchies, but the
+specification is very complex and parsing YAML is considered risky due to
+potential of injection vulnerabilities that is common when trying to support
+dynamic local object creation as defined by the specification. The application
+configuration is in TOML format, so having to deal with multiple serialisation
+formats when creating an application is not a great developer experience. Is
+there a suitable format that will work for everything?
+
+### Invocation Routing
+
+The planner node will be responsible for creating the required pseudo nodes as
+well. In the specific cases of HTTP- and gRPC pseudo nodes each pseudo node will
+send all invocations over a single dedicated invocation channel. The planner
+node will, by default, have the handle to receive these invocations and
+therefore be responsible for routing them to the right Wasm nodes based on the
+associated labels (specifically the label associated with the request receiver
+channel should match the label of the target Wasm node). Since it would be
+possible for multiple Wasm nodes to share the same label, the planner node would
+need an indication of which Wasm node should receive invocations with a specific
+label on the request receiver channel.
```

Comment:
Who will indicate what node is the destination? Will it be a new value in the gRPC metadata?

Reply:
The only thing that the router node can read (without the need to have at least the same label as the target) is the label on the invocation. This means the label is the only thing that the router can use to make its decision.